File: civitasjava/src/Civitas/Jugador.java
package civitas;

import static java.lang.Float.compare;
import java.util.ArrayList;
import civitas.TituloPropiedad;
import civitas.Casilla;

import java.lang.Comparable;

public class Jugador implements Comparable<Jugador>{
   
    static protected int CasasMax = 4;
    
    static protected int CasasPorHotel = 4;
    
    protected boolean encarcelado;
    
    static protected int HotelesMax = 4;
    
    private String nombre;
    
    private int numCasillaActual;
    
    static protected float PasoPorSalida = 1000f;
    
    static protected float PrecioLibertad = 200f;
    
    private boolean puedeComprar;
    
    protected float saldo;
    
    static float SaldoInicial = 7500f;

    protected SorpresaSalirCarcel salvoconducto = null;
    
    protected ArrayList<TituloPropiedad> propiedades;
    
    protected Boolean especulador;
    
    
    
    public Boolean getEspeculador(){
        return especulador;
    }
    
    @Override
    public String toString(){
        return "Jugador{ Encarcelado: " +encarcelado +", nombre: " +nombre 
                +", numCasillaActual: " +numCasillaActual+ ", puedeComprar=" + 
                puedeComprar + ", saldo=" + saldo + ", propiedades=" + 
                propiedades.toString() + ", salvoconducto=" + salvoconducto 
                + '}'+"\n";
    }
    /**
     * Sirve para cancelar la hipoteca de la propiedad pasada como argumento
     * @param ip : indice de la propiedad
     * @return res
     */
    boolean cancelarHipoteca(int ip) {
        boolean res = false;
        if(encarcelado == true){
            return res;
        }
        
        if(existeLaPropiedad(ip)){
            TituloPropiedad propiedad = propiedades.get(ip);
            float cantidad = propiedad.getImporteCancelarHipoteca();
            boolean puedoGastar = puedoGastar(cantidad);
        
            if(puedoGastar){
                res = propiedad.cancelarHipoteca(this);
                if(res){
                    Diario.getInstance().ocurreEvento("El jugador " +nombre+ 
                        "cancela la hipoteca de la propiedad " +ip);
                }
            }
        }
        return res;
    }
    
    /**
     * Calcula la cantidad de casas y hoteles
     * @return cantidad
     */
    int cantidadCasasHoteles() {
        int cantidad = 0;
        
        for(int i=0; i<propiedades.size(); i++){
            cantidad+=propiedades.get(i).getNumCasas();
            cantidad+=propiedades.get(i).getNumHoteles();
        }
        return cantidad;
    }
    
    /**
     * Compara dos jugadores por su saldo
     * @param otro : otro jugador para comparar
     * @return 
     */
    
    @Override
    public int compareTo(Jugador otro) {
        return compare(saldo, otro.getSaldo());
    }
    
    /**
     * Compra una propiedad
     * @param titulo : titulo de la propiedad
     * @return res
     */
    boolean comprar(TituloPropiedad titulo) {
        boolean res = false;
        
        if(encarcelado){
            return res;
        }
        
        if(puedeComprar){
            float precio = titulo.getPrecioCompra();
            
            if(puedoGastar(precio)){
                res = titulo.comprar(this);
                
                if(res){
                    propiedades.add(titulo);
                    Diario.getInstance().ocurreEvento("El jugador "+this 
                    +" compra la propiedad " +titulo.toString());
                    
                }
                puedeComprar = false;
            }
        }
        return res;
    }
    
    /**
     * Construye casa en la propiedad indicada como argumento
     * @param ip : indice de la propiedad
     * @return res
     */
    boolean construirCasa(int ip) {
        boolean res = false;
        
        if(encarcelado){
            return res;
        }
        else{
            if(existeLaPropiedad(ip)){
                TituloPropiedad propiedad = propiedades.get(ip);
                boolean puedoEdificarCasa = puedoEdificarCasa(propiedad);
                
                if(puedoEdificarCasa){
                    res = propiedad.construirCasa(this);
                    if(res){
                        Diario.getInstance().ocurreEvento("El jugador "+nombre 
                        +" construye casa en la propiedad " +
                                propiedades.get(ip).getNombre());
                    }
                }   
            }
        }
        
        return res;
    }
    
    /**
     * Construye hotel en la porpiedad indicada como argumento
     * @param ip : indice de la propiedad
     * @return res
     */
    boolean construirHotel(int ip) {
        boolean res = false;
        if(encarcelado){
            return res;
        }
        
        if(existeLaPropiedad(ip)){
            TituloPropiedad propiedad = propiedades.get(ip);
            boolean puedoEdificarHotel = puedoEdificarHotel(propiedad);
            float precio = propiedad.getPrecioEdificar();
            
            if(puedoGastar(precio) && propiedad.getNumHoteles()<getHotelesMax()
                    && propiedad.getNumCasas() >= getCasasPorHotel()){
                puedoEdificarHotel = true;
            }
            
            if(puedoEdificarHotel){
                res = propiedad.construirHotel(this);
                propiedad.derruirCasas(CasasPorHotel, this);
            }
            
            if(res){
                Diario.getInstance().ocurreEvento("El jugador " +nombre 
                +"construye hotel en la propiedad " +
                        propiedades.get(ip).getNombre());
            }
        }
        
        return res;
    }
    
    /**
     * Devuelve true si el jugador debe ser encarcelado
     * @return res
     */
    protected boolean debeSerEncarcelado() {
        boolean res;
        if(isEncarcelado()){
            res = false;
        }else if(tieneSalvoconducto()){
            perderSalvoconducto();
            res = false;
            Diario.getInstance().ocurreEvento("El jugador" +nombre +"ha perdido"
                + " salvoconducto");
        }else{
            res = true;
        }
        
        return res;
    }
    
    /**
     * Muestra los nombres de las propiedades
     * @return nombres
     */
    public ArrayList<String> getNombrePropiedades(){
        ArrayList<String> nombres = new ArrayList<>();      
        for(int i=0;i<propiedades.size();i++)
            nombres.add(propiedades.get(i).getNombre());
        return nombres;
       
    }
    
    
    /**
     * Devuelve true si el jugador esta en bancarrota
     * @return saldo
     */
    boolean enBancarrota() {
        return saldo <= 0;
    }
    
    /**
     * Encarcela al jugador si debe ser encarcelado y se informa
     * @param numCasillaCarcel : casilla carcel
     * @return encarcelado
     */
    boolean encarcelar(int numCasillaCarcel) {
        if(debeSerEncarcelado()){
            moverACasilla(numCasillaCarcel);
            encarcelado = true;
        }
        return encarcelado;
    }
    
    protected boolean existeLaPropiedad(int ip) {
        return (ip >= 0 && ip < propiedades.size());
    }
    
    protected int getCasasMax() {
        return CasasMax;
    }
    
    int getCasasPorHotel() {
        return CasasPorHotel;
    }
    
    protected int getHotelesMax() {
        return HotelesMax;
    }
    
    public String getNombre() {
        return nombre;
    }
    
    int getNumCasillaActual() {
        return numCasillaActual;
    }
    
    private float getPrecioLibertad() {
        return PrecioLibertad;
    }
    
    private float getPrecioPasoSalida() {
        return PasoPorSalida;
    }
    
    public ArrayList<TituloPropiedad> getPropiedades() {
        return propiedades;
    }
    
    boolean getPuedeComprar() {
        return puedeComprar;
    }
    
    public float getSaldo() {
        return saldo;
    }
    
    Sorpresa getSalvoconducto(){
        return salvoconducto;
    }
    
    /**
     * Hipoteca la propiedad pasada como parametro
     * @param ip : indice de la propiedad
     * @return res
     */
    boolean hipotecar(int ip) {
        boolean res = false;
        
        if(encarcelado){
            return res;
        }
        
        if(existeLaPropiedad(ip)){
            TituloPropiedad propiedad = propiedades.get(ip);
            res = propiedad.hipotecar(this);
        }
        
        if(res){
            Diario.getInstance().ocurreEvento("El jugador " +nombre 
            +"hipoteca la propiedad " +ip);
        }
        return res;
    }
    
    public Boolean isEncarcelado() {
        return encarcelado;
    }
    
    Jugador(String nombre) {
        this.nombre = nombre;
        propiedades = new ArrayList<>();
        saldo = SaldoInicial;
        numCasillaActual = 0;
        puedeComprar = true;
    }
    
    protected Jugador(Jugador otro) {
        encarcelado = otro.isEncarcelado();
        nombre = otro.getNombre();
        numCasillaActual = otro.getNumCasillaActual();
        puedeComprar = otro.getPuedeComprar();
        saldo = otro.getSaldo();
        salvoconducto = otro.salvoconducto;
        propiedades = otro.getPropiedades();
    }
    
    /**
     * Modifica el saldo del jugador sumandole (cantidad)
     * @param cantidad : cantidad a incrementar
     * @return 
     */
    Boolean modificarSaldo(float cantidad) {
        saldo += (float)(cantidad);
        Diario.getInstance().ocurreEvento("Se ha modificado el saldo del "
                + "jugador " +nombre);
        return true;
    }
    
    /**
     * Se mueve el jugador a la casilla pasada como parametro 
     * si no esta encarcelado
     * @param numCasilla : casilla a la que se mueve
     * @return res
     */
    boolean moverACasilla(int numCasilla) {
        boolean res;
        if(isEncarcelado()){
            res = false;
        }
        else{
            numCasillaActual = numCasilla;
            puedeComprar = false;
            Diario.getInstance().ocurreEvento("El jugador " +nombre 
                    +" se mueve de la casilla " +numCasillaActual +
                    " a la " +numCasilla);
            res = true;
        }
        
        return res;
    }
    
    /** 
     * Si el jugador no esta encarcelado se devuelve true
     * @param sorpresa :  
     * @return res
     */
    boolean obtenerSalvoconducto(SorpresaSalirCarcel sorpresa) {
        boolean res;
        if(encarcelado){
            res = false;
        }
        else{
            salvoconducto = sorpresa;
            res = true;
        }
        return res;
    }
    
    /**
     * El jugador paga una cantidad
     * @param cantidad : cantidad que tiene que pagar
     * @return
     */
    boolean paga(float cantidad) {
        return modificarSaldo(cantidad*(-1));
    }
    
    /**
     * El jugador pague el alquiler si no esta encarcelado
     * @param cantidad : cantidad que tiene que pagar
     * @return res
     */
    boolean pagaAlquiler(float cantidad) {
        boolean res = false;
        if(encarcelado){
            res = false;
        }
        else{
            res = paga(cantidad);
        }
        
        return res;
    }

    /**
     * Metodo que hace que el jugador pague impuesto si no esta encarcelado
     * @param cantidad : cantidad que tiene que pagar
     * @return res
     */
    boolean pagaImpuesto(float cantidad) {
        boolean res;
        
        if(encarcelado){
            res = false;
        }
        else{
            res = paga(cantidad);
        }
        
        return res;
    }
    /**
     * Jugador pasa por la salida y se le cobra. Se informa en el Diario
     * @return
     */
    boolean pasaPorSalida() {
        modificarSaldo(PasoPorSalida);
        Diario.getInstance().ocurreEvento("El jugador " +nombre +"ha pasado por"
                + "la salida");
        return true;
    }
    
    /**
     * El jugador pierde salvoconducto
     */
    private void perderSalvoconducto() {
        salvoconducto.usada();
        salvoconducto = null;
    }
    
    /**
     * Consulta si el jugador puede comprar unacasilla
     * @return 
     */
    boolean puedeComprarCasilla() {
        puedeComprar = true;
        if(encarcelado){
            puedeComprar = false;
        }
        
        return puedeComprar;
    }
    
    /**
     * Si el jugador esta en la carcel y puede salir pagando, se paga el precio,
     * deja de estar encarcelado y se informa al diario
     * @return resultado
     */
    boolean salirCarcelPagando() {
        boolean resultado = false;

        if(encarcelado && puedesalirCarcelPagando()){
            paga(PrecioLibertad);
            encarcelado = false;
            Diario.getInstance().ocurreEvento(nombre +" ha pagado para salir "
                    + "de la carcel");
            resultado = true;
        }

        return resultado;
    }
    
    /**
     * Informa de si se puede edificar y en caso de que se puede cual seria
     * el precio de esto
     * @param propiedad : propiedad en la que se quiere edificar
     * @return res
     */
    private boolean puedoEdificarCasa(TituloPropiedad propiedad) {
        boolean res = false;
        if(encarcelado == false){
            if(propiedad.getNumCasas()<=CasasMax){
                res = saldo >= propiedad.getPrecioEdificar();
            }
        }
        return res;
    }
    
    /**
     * Informa de si se puede edificar un hotel y en caso de que se pueda
     * cual seria el precio de este
     * @param propiedad : propiedad en la que se quiere edificar
     * @return res
     */
    private boolean puedoEdificarHotel(TituloPropiedad propiedad) {
        boolean res = false;
        if(encarcelado == false){
            if(propiedad.getNumCasas()<=HotelesMax && 
                    propiedad.getNumCasas()==CasasMax){
                res = saldo >= propiedad.getPrecioEdificar();
            }
        }
        
        return res;
    
    }
    
    /**
     * Consulta si el jugador puede gastarse ese dinero
     * @param precio : cantidad que quiero comprobar
     * @return res
     */
    protected boolean puedoGastar(float precio) {
        boolean res = false;
        if(encarcelado){
            res = false;
        }else{
            if(saldo >= precio){
                res = true;
            }
        }
        
        return res;
    }
    
    /**
     * Hace recibir al jugador una cantidad de dinero
     * @param cantidad : cantidad de dinero que recibe
     * @return res
     */
    boolean recibe(float cantidad) {
        boolean res;
        if(encarcelado){
            res = false;
        }else{
            res = modificarSaldo(cantidad);
        }
        
        return res;
    }
    
    /**
     * Informa si el saldo del jugador es mayor al precio de salir de la carcel
     * @return 
     */
    boolean puedesalirCarcelPagando() {
        return (saldo >= PrecioLibertad);        
    }
    
    /**
     * Se pregunta al dado si el jugador debe salir de la carcel
     * @return res
     */
    boolean salirCarcelTirando() {
        boolean res = false;
        
        if(Dado.getInstance().salgoDeLaCarcel()){
            encarcelado = false;
            Diario.getInstance().ocurreEvento(nombre +"ha tirado para salir "
                    + "de la carcel"); 
            res = true;
        }
        return res;
    }
    
    /**
     * Indica si el jugador tiene propiedades
     * @return 
     */
    boolean tieneAlgoQueGestionar() {
        return (propiedades!=null);
    }
    /**
     * Indica si el jugador tiene un Salvoconducto
     * @return 
     */
    boolean tieneSalvoconducto() {
        return (salvoconducto!=null);
    }
    
    /**
     * Muestra informacion
     * @return resultado
     */
    /*public String toString() {
        String resultado ="";
        String listapropiedades = "";
        String aux;
        
        for(int i=0; i<propiedades.size(); i++){
            aux = propiedades.get(i).getNombre();
            listapropiedades += " ";
            listapropiedades += aux;
        }
        
        resultado = "" +nombre +"\nEncarcelado: " +encarcelado + 
                "\nPuedecomprar: " +puedeComprar +"\nSaldo: " +saldo
                +"\nNº propiedades: " +propiedades.size()+ "\nPropiedades" 
                +propiedades +"\nNº casilla actual: " +numCasillaActual;
        
        return resultado;
    };*/
    
    /**
     * Se vende la propiedad (ip)
     * @param ip : propiedad que se quiere vender
     * @return resultado
     */
    boolean vender(int ip){
        boolean resultado = false;
        if(isEncarcelado()){
            resultado = false;
        }
        else{
            if(existeLaPropiedad(ip)){
                resultado = propiedades.get(ip).vender(this);
                if(resultado){
                    Diario.getInstance().ocurreEvento(nombre +"ha vendido la"
                        +" casilla " +propiedades.get(ip));
                    propiedades.remove(ip);
                }                
            }
        }
        
        return resultado;
    }

    void setPuedeComprar(boolean d){
        puedeComprar = d;
    }
}

File: RubyCivitas/pkg/RubyApplication1-0.0.1/lib/jugador.rb
require_relative "diario"
require_relative "titulo_propiedad"
module Civitas
  class Jugador

    @@casas_max = 4
    @@casas_por_hotel = 4
    @@hoteles_max = 4
    @@paso_por_salida = 1000
    @@precio_libertad = 200
    @@saldo_inicial = 7500

    def initialize(nombre)
      @encarcelado = false
      @nombre = nombre
      @num_casilla_actual = 1
      @puede_comprar = true
      @saldo = @@saldo_inicial
      @salvoconducto = false
      @propiedades = Array.new
    end

    def debe_ser_encarcelado

      if !@encarcelado && !tiene_salvoconducto
        return true
      elsif !@encarcelado && tiene_salvoconducto
        perder_salvoconducto
        Diario.instance.ocurre_evento("[DIARIO] El jugador #{@nombre} se libra de la carcel.")
        return false
      else
        return true
      end
      return false
    end

    def encarcelar(num_casilla_carcel)

      if debe_ser_encarcelado

        mover_a_casilla(num_casilla_carcel)
        @encarcelado = true
        Diario.instance.ocurre_evento("[DIARIO] El jugador #{@nombre} ha sido encarcelado")
      end

      return @encarcelado
    end

    def existe_la_propiedad(ip)

      return (@propiedades.size > ip) && (ip >= 0)

    end

    def obtener_salvoconducto(s)

      if @encarcelado
        return false
      else
        @salvoconducto = true
      end
      @salvoconducto
    end

    def perder_salvoconducto

      usada
      @salvoconducto = nil

    end

    def tiene_salvoconducto

      !@salvoconducto.nil?

    end

    def puede_comprar_casilla

      if @encarcelado

        @puede_comprar = false
      else

        @puede_comprar = true

      end

      @puede_comprar
    end

    def paga(cantidad)

      modificar_saldo(cantidad * -1)

    end

    def paga_impuesto(cantidad)

      if @encarcelado
        false
      else
        paga(cantidad)
      end
    end

    def paga_alquiler(cantidad)

      if @encarcelado
        false
      else
        paga(cantidad)
      end

    end

    def recibe(cantidad)

      if @encarcelado
        false
      else
        modificar_saldo(cantidad)
      end
    end

    def modificar_saldo(cantidad)

      @saldo += cantidad
      Diario.instance.ocurre_evento("[DIARIO] El saldo del jugador ha sido incrementado en " + cantidad.to_s)
      return true
    end

    def mover_a_casilla(num_casilla)

      if @encarcelado

        false
      else
        @num_casilla_actual = num_casilla
        @puede_comprar = false
        Diario.instance.ocurre_evento("[DIARIO] El jugador ha sido movido a la casilla " + num_casilla.to_s)
        true
      end
    end

    def puedo_gastar(precio)

      if @encarcelado
        false
      else
        @saldo >= precio
      end

    end

    def vender(ip)

      if @encarcelado
        false
      elsif existe_la_propiedad(@propiedades.at(ip))

        @propiedades.at(ip).vender(self)
        @propiedades.delete(@propiedades.at(ip))

        true
      end
    end

    def tiene_algo_que_gestionar

      return @propiedades.size >= 1

    end

    def puede_salir_carcel_pagando

      @saldo >= @@precio_libertad

    end

    def salir_carcel_pagando

      if @encarcelado && puede_salir_carcel_pagando

        paga(@@precio_libertad)
        @encarcelado = false
        Diario.instance.ocurre_evento("[DIARIO] El jugador #{@nombre} ha salido de la carcel")
        true
      else
        false
      end

    end

    def salir_carcel_tirando

      Dado.instance.salgo_de_la_carcel

    end

    def pasa_por_salida

      modificar_saldo(@@paso_por_salida)
      Diario.instance.ocurre_evento("[DIARIO] El jugador #{@nombre} ha pasado por salida +#{@@paso_por_salida}")
      true

    end

    def puedo_edificar_hotel(propiedad)

      @propiedades.contains(propiedad) && puedo_gastar(propiedad.precio_edificar) &&
        (propiedad.num_casas >= @@casas_por_hotel) && (propiedad.num_hoteles < @@hoteles_max)

    end

    def puedo_edificar_casa(propiedad)

      @propiedades.contains(propiedad) && puedo_gastar(propiedad.precio_edificar) &&
        propiedad.num_casas < @@casas_max

    end

    def compare_to(otro = Jugador)

      @saldo <=> otro.saldo

    end

    def cancelar_hipoteca(ip)

      boolean result = false

      if !@encarcelado && existe_la_propiedad(ip)

        propiedad = @propiedades.at(ip)
        cantidad = @propiedades.get_importe_cancelar_hipoteca

        if puedo_gastar(cantidad)

          result = propiedad.cancelar_hipoteca(self)

        end

        if result

          Diario.instance.ocurre_evento("[DIARIO] El jugador " + @nombre + " cancela la hipoteca de la propiedad " + ip.to_s)

        end
      end

      result
    end

    def comprar(titulo)

      result = false;

      if !@encarcelado && @puede_comprar

        precio_compra = titulo.precio_compra

        if puedo_gastar(precio_compra)

          @propiedades.push(titulo)
          result = titulo.comprar(self)

          if result

            @propiedades.push(titulo)
            Diario.instance.ocurre_evento("[DIARIO] El jugador " + @nombre + " compra " + titulo.nombre)

          end

        end

        @puede_comprar = false

      end

      result

    end

    def construir_hotel(ip)

      if !@encarcelado && existe_la_propiedad(ip)

        propiedad = @propiedades.at(ip)
        puedo_edificar_hotel = puedo_edificar_hotel(propiedad)
        precio = propiedad.precio_edificar

        if puedo_edificar_hotel

          result = propiedad.construir_hotel(self)

          casas_por_hotel = @@casas_por_hotel
          propiedad.derruir_casas(casas_por_hotel, self)

          Diario.instance.ocurre_evento("[DIARIO] El jugador " + @nombre + " construye un hotel " + "en " +
                                     ip + ", " + propiedad.nombre)
        end
      end

      result
    end

    def construir_casa(ip)

      result = false

      if ip >= 0 && ip < @propiedades.size

        propiedad = @propiedades.at(ip)

        if !@encarcelado && existe_la_propiedad(ip)

          puedo_edificar_casa = puedo_edificar_casa(propiedad)
          precio = propiedad.precio_edificar

          if (puedo_edificar_casa)

            result = propiedad.construir_casa(self)

          end
        end

        if result

          Diario.instance.ocurre_evento("[DIARIO] El jugador #{@nombre} construye una casa en la propiedad #{ip}")

        end

      end
    end

    def hipotecar(ip)

      result = false

      if !@encarcelado && existe_la_propiedad(ip)

        propiedad = @propiedades.at(ip)
        result = propiedad.hipotecar(self)

      end

      if result

        Diario.instance.ocurre_evento("[DIARIO] El jugador #{@nombre} hipoteca la propiedad #{ip}")

      end

      result
    end

    def is_encarcelado

      @encarcelado

    end

    def to_string

      "-----------[#{@nombre}]------------\n[Saldo] -> #{@saldo}\n[Encarcelado] -> #{@encarcelado}\n[Numero de casilla actual] -> #{@num_casilla_actual}\n"

    end
    
    public

    attr_reader :casas_max, :casas_por_hotel, :hoteles_max, :nombre, :precio_libertad,
      :paso_por_salida, :propiedades, :puede_comprar, :saldo
    
    attr_accessor :num_casilla_actual

    protected :debe_ser_encarcelado

    private :puede_salir_carcel_pagando, :puedo_gastar, :perder_salvoconducto

    public :compare_to, :vender

  end
end
